File: continuum-idl/src/main/java/org/kinotic/continuum/idl/api/schema/FunctionDefinition.java
/*
 *
 * Copyright 2008-2021 Kinotic and the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.kinotic.continuum.idl.api.schema;

import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import lombok.*;
import lombok.experimental.Accessors;
import org.apache.commons.lang3.Validate;
import org.kinotic.continuum.idl.api.schema.decorators.C3Decorator;

import java.util.LinkedList;
import java.util.List;

/**
 * Provides functionality to define a function with a Continuum schema.
 * The context for equality here is the {@link ServiceDefinition}.
 * Given that no two functions can have the same name in the same {@link ServiceDefinition}.
 * Created by navid on 2023-4-13
 */
@Getter
@Setter
@Accessors(chain = true)
@NoArgsConstructor
@AllArgsConstructor
@EqualsAndHashCode(callSuper = false)
@ToString
public class FunctionDefinition extends AbstractDefinition {

    /**
     * The name of this {@link FunctionDefinition}
     */
    private String name;

    /**
     * This is the {@link C3Type} that defines the return type of this function.
     */
    @EqualsAndHashCode.Exclude
    private C3Type returnType = new VoidC3Type();

    /**
     * The list of arguments that this function takes
     */
    @JsonDeserialize(as = LinkedList.class)
    private LinkedList<ParameterDefinition> parameters = new LinkedList<>();

    /**
     * Adds a new parameter to this function
     * @param name the name of the parameter
     * @param c3Type the type of the parameter
     * @return this {@link FunctionDefinition} for chaining
     */
    public FunctionDefinition addParameter(String name, C3Type c3Type){
        ParameterDefinition parameter = new ParameterDefinition(name, c3Type);
        return addParameter(parameter);
    }

    /**
     * Adds a new argument to this function
     * @param name the name of the parameter
     * @param c3Type the type of the parameter
     * @param decorators the decorators to apply to the parameter
     * @return this {@link FunctionDefinition} for chaining
     */
    public FunctionDefinition addParameter(String name, C3Type c3Type, List<C3Decorator> decorators){
        ParameterDefinition parameter = new ParameterDefinition(name, c3Type);
        parameter.setDecorators(decorators);
        return addParameter(parameter);
    }

    /**
     * Adds a new parameter to this function
     * @param parameter the parameter to add
     * @return this {@link FunctionDefinition} for chaining
     */
    public FunctionDefinition addParameter(ParameterDefinition parameter){
        Validate.isTrue(!parameters.contains(parameter), "FunctionDefinition already contains parameter "+parameter.getName());
        parameters.add(parameter);
        return this;
    }

}

File: continuum-js/continuum-idl-js/src/api/FunctionDefinition.ts
import {AbstractDefinition} from '@/api/AbstractDefinition'
import {ParameterDefinition} from '@/api/ParameterDefinition'
import { C3Type } from '@/api/C3Type'
import {C3Decorator} from '@/api/decorators/C3Decorator'
import {MetadataType} from '@/api/HasMetadata'
import { VoidC3Type } from '@/api/VoidC3Type'

/**
 * Provides functionality to define a function with a Continuum schema.
 * Created by navid on 2023-4-13
 */
export class FunctionDefinition extends AbstractDefinition {

    /**
     * This is the C3Type that defines the return type of this function.
     */
    public returnType: C3Type = new VoidC3Type()

    /**
     * This array defines the {@link C3Type}s that define the arguments for this function.
     * Argument names must be unique.
     */
    public parameters: ParameterDefinition[] = []


    constructor(name: string,
                decorators?: C3Decorator[] | null,
                metadata?: MetadataType | null) {
        super(name, decorators, metadata)
    }

    /**
     * Adds a new {@link ParameterDefinition} to this {@link FunctionDefinition}
     * {@link ParameterDefinition} must have unique names
     * @param name the name of the parameter
     * @param c3Type the type of the parameter
     * @param decorators the decorators to apply to the parameter
     */
    public addParameter(name: string, c3Type: C3Type, decorators?: C3Decorator[]): FunctionDefinition {
        const param = new ParameterDefinition(name, c3Type, decorators)
        if(this.parameters.find((value) => value.name === name)){
            throw new Error(`FunctionDefinition already contains parameter for name ${name}`);
        }
        this.parameters.push(param)
        return this
    }

}
